### PR TITLE
gh-114685: `PyBuffer_FillInfo` now raises on `PyBUF_{READ,WRITE}`

### PR DESCRIPTION
**Pull request by sobolevn**
* Issue: gh-114685

### CODE REVIEW

File: Modules/_testcapimodule.c [hunk 1]
```diff
@@ -1261,6 +1261,25 @@ make_memoryview_from_NULL_pointer(PyObject *self, PyObject *Py_UNUSED(ignored))
...
+    int readonly;
+    int flags;
+
+    if (!PyArg_ParseTuple(args, "s#ii:buffer_fill_info",
+                          &data, &size, &readonly, &flags)) {
+        return NULL;
+    }
+
+    if (PyBuffer_FillInfo(&info, NULL, (void *)data, size, readonly, flags) < 0)
+        return NULL;
```
Comment (aisk) [inline Modules/_testcapimodule.c]:
```suggestion
    if (PyBuffer_FillInfo(&info, NULL, (void *)data, size, readonly, flags) < 0) {
        return NULL;
    }
```

File: Lib/test/test_buffer.py [hunk 2]
```diff
@@ -4591,6 +4591,23 @@ def test_c_buffer_invalid_flags(self):
...
+    @unittest.skipIf(_testcapi is None, "requires _testcapi")
+    def test_c_fill_buffer_invalid_flags(self):
+        # PyBuffer_FillInfo
+        self.assertRaises(SystemError, _testcapi.buffer_fill_info,
+                          "abc", 0, PyBUF_READ)
+        self.assertRaises(SystemError, _testcapi.buffer_fill_info,
+                          "abc", 0, PyBUF_WRITE)
+
+    @unittest.skipIf(_testcapi is None, "requires _testcapi")
+    def test_c_fill_buffer_readonly_and_writable(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_buffer.py:4604]: See #60198. `PyBuffer_FillInfo()` with NULL as exporter is a fragile thing. In your test you depend on the GC to release the memoryview. At least release it deterministically (by calling `.release()` or using `with`).

File: Lib/test/test_buffer.py [hunk 3]
```diff
@@ -4591,6 +4591,23 @@ def test_c_buffer_invalid_flags(self):
...
+        self.assertEqual(
+            str(_testcapi.buffer_fill_info("abc", 1, PyBUF_SIMPLE), "utf8"),
```
Comment (serhiy-storchaka) [inline Lib/test/test_buffer.py]: Why not use `bytes()`.

You can also test it with PyBUF_WRITABLE flag and readonly=0. But you need to fill the buffer from a bytearray for safety. Test also that the resulting memoryview is actually readonly or writeable.

**Review by serhiy-storchaka — COMMENTED** [review]
See #60198. `PyBuffer_FillInfo()` with NULL as exporter is a fragile thing.

**Review by serhiy-storchaka — COMMENTED** [review]
Thank you.

Yet one nitpick. The code relies on the fact that `b"abc"` still exists after returning from `buffer_fill_info()`. It is so in CPython, but this is an implementation detail. It would be more reliable to explicitly keep a reference to it:
```py
source = b'abc'
...
buffer_fill_info(source, ...)
```

**Comment by sobolevn** [comment]
Thank *you*! I always learn a lot for your reviews, very helpful.

File: Lib/test/test_buffer.py [hunk 4]
```diff
@@ -4591,6 +4591,26 @@ def test_c_buffer_invalid_flags(self):
...
+    def test_c_fill_buffer_readonly_and_writable(self):
+        source = b"abc"
+        with _testcapi.buffer_fill_info(source, 1, PyBUF_SIMPLE) as m:
+            self.assertEqual(bytes(m), b"abc")
+            self.assertTrue(m.readonly)
+        with _testcapi.buffer_fill_info(source, 0, PyBUF_WRITABLE) as m:
+            self.assertEqual(bytes(m), b"abc")
+            self.assertFalse(m.readonly)
+        self.assertRaises(BufferError, _testcapi.buffer_fill_info,
+                          "abc", 1, PyBUF_WRITABLE)
```
Comment (serhiy-storchaka) [inline Lib/test/test_buffer.py]:
```suggestion
        self.assertRaises(BufferError, _testcapi.buffer_fill_info,
                          source, 1, PyBUF_WRITABLE)
```

File: Lib/test/test_buffer.py [hunk 5]
```diff
@@ -4591,6 +4591,26 @@ def test_c_buffer_invalid_flags(self):
         self.assertRaises(SystemError, buf.__buffer__, PyBUF_READ)
         self.assertRaises(SystemError, buf.__buffer__, PyBUF_WRITE)
 
+    @unittest.skipIf(_testcapi is None, "requires _testcapi")
+    def test_c_fill_buffer_invalid_flags(self):
+        # PyBuffer_FillInfo
+        self.assertRaises(SystemError, _testcapi.buffer_fill_info,
+                          "abc", 0, PyBUF_READ)
+        self.assertRaises(SystemError, _testcapi.buffer_fill_info,
+                          "abc", 0, PyBUF_WRITE)
```
Comment (serhiy-storchaka) [inline Lib/test/test_buffer.py]:
```suggestion
        source = b"abc"
        self.assertRaises(SystemError, _testcapi.buffer_fill_info,
                          source, 0, PyBUF_READ)
        self.assertRaises(SystemError, _testcapi.buffer_fill_info,
                          source, 0, PyBUF_WRITE)
```

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Pull request merged**